text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+
+Note that "any type which has a field which implements `poll_drop_ready` means,
+in practice, *any generic type*, because any generic parameter could possibly
+implement `poll_drop_ready`. This would be incredibly burdensome across the
+ecosystem.
+
+## Have no virtual async drop, more weirdness around auto traits
+
+The ideal and most immediately obvious API for an async destructor is something
+like this:
+
+```rust
+trait AsyncDrop {
+    async fn drop(&mut self);
+}
+```
+
+However, this introduces several problems because it would allow the body of
+the async fn to introduce arbitrary additional state not contained in the type
+itself. This would cause the same problems that non-fused semantics would cause
+for drop glue, but it would also cause even more problems.
+
+The first is that we would not be able to async drop a trait object. When a
+trait object is dropped, we dynamically dispatch the call to the destructor of
+the concrete type that backs that object, through the vtable of the trait
+object. However, this definition of async drop would require some place to put
+the state of the future the async call returns, and the layout of each future
+for each concrete type would be different and unknown. It would be functionally
+equivalent to having a trait object with an unknown associated type, which is
+not allowed for exactly this reason. It would not be possible to support async
+destructors on trait objects.
+
+It would also make the problem of futures created with async contexts not
+implementing `Send` or `Sync` worse. If the state of the future returned by the
+async drop was not `Send` or `Sync`, neither would any async context in which
+that type is dropped be `Send` or `Sync`.
+
+In other words, even though the user cannot call the async drop method of a
+type, using an async fn for this purpose would introduce many of the
+limitations of async fn in traits, and would not be ideal for the use case of
+destructors.
+
+## Putting `poll_drop_ready` in a separate trait
+
+Possibly, `poll_drop_ready` should not be a provided method of Drop, but should
+be in a separate trait. This would increase the complexity of the API, but make
+it possible to bound by that trait. In general, bounding by Drop is usually an
+antipattern that doesn't do what the user wants and there's not really any use
+in creating a separate trait for this method either.
+
+# Prior art
+[prior-art]: #prior-art
+
+No other language seems to have a way to call asynchronous code from the
+destructor without spawning a new task. There is one major factor that most of
+those languages have in common: they are garbage collected, and so their
+equivalent to destructors (usually called finalizers) cannot be easily
+guaranteed to run at any particular time. For that reason, they are not used
+for clean up of external resources in the "RAII" sense that they are used in
+Rust and C++.
+
+C++ has a coroutine proposal similar to async/await. It has not yet been
```

Comment:
I think this is a wrong description of it. It doesn't need async destructors because C++ coroutines which resemble async/await tasks run to completion. You can just write your normal function, with a `try/finally` block if you desire some cleanup.

See my old question to Lewis Baker about this: https://github.com/lewissbaker/cppcoro/issues/102

The same applies e.g. to async/await in Javascript, Kotlin, C#, etc.
The Rust situation is a homemade issue, since it's the only language which represents async functions as futures which are cancellable at any suspension point.